pycloudlib/qemu/cloud.py

```diff
@@ -0,0 +1,651 @@
+                relative to the image_dir.
+            instance_type: Type of instance to create. For QEMU, this is
+                in the form of c#m#, where # is the number of cpus and
+                memory in MB.
+            user_data: used by cloud-init to run custom scripts/configuration
+            meta_data: used by cloud-init for custom metadata
+            vendor_data: used by cloud-init for custom vendor data
+            kernel_cmdline: kernel command line arguments
+            kernel_path: path to kernel to use
+            no_seed_iso: if True, do not create a seed iso
+            username: username to use for ssh connection
+            launch_args: list of additional arguments to pass to qemu
+
+
+        Returns:
+            An instance object to use to manipulate the instance further.
+
+        """
+        # Start with early validation of parameters
+        base_image = self._find_base_image(image_id=image_id)
+        cpus, memory = self._parse_instance_type(instance_type=instance_type)
+
+        # Next create the dir to contain all of the instance artifacts
+        instance_dir = self._get_available_file(
+            self.parent_dir / Path(image_id).stem
+        )
+        instance_dir.mkdir()
+        self._log.info(
+            "Using instance dir '%s' for new instance launched from '%s'",
+            instance_dir,
+            image_id,
+        )
+
+        # We make a QCOW image from the base image so we can make
+        # destructive changes without modifying the original image
+        instance_path = instance_dir / "inst.qcow2"
+        self._create_qcow_image(
+            instance_path=instance_path, base_image=base_image
+        )
+
+        ssh_port = get_free_port()
+        telnet_port = get_free_port()
+        socket_path = instance_dir / "qmp-socket"
+        qemu_args = [
+            self.qemu_binary,
+            "-enable-kvm",
+            "-cpu",
+            "host",
+            "-smp",
+            f"cpus={cpus}",
+            "-m",
+            f"size={memory}",
+            "-net",
+            "nic",
+            "-net",
+            f"user,hostfwd=tcp::{ssh_port}-:22",
+            "-hda",
+            str(instance_path),
+            "-nographic",
+            "-chardev",
+            f"socket,id=char0,logfile={instance_dir / 'console.log'},"
+            f"host=0.0.0.0,port={telnet_port},telnet=on,server=on,"
+            "wait=off,mux=on",
+            "-serial",
+            "chardev:char0",
+            "-qmp",
+            f"unix:{socket_path},server,wait=off",
+            "-no-shutdown",
+        ]
+
+        if not no_seed_iso:
+            # By default, even if no user data is passed, we'll need to
+            # use NoCloud and create a seed iso, otherwise we have no
+            # way of accessing our instance
+            seed_path = self._create_seed_iso(
+                instance_dir=instance_dir,
```

Comment:
Do we want the opportunity to provide network_config param at launch and inject it into the seed ISO as well for NoCloud datasource? I can think of more complex cases where we may want to add mutiple nics and or setup bridges/bonds etc and validate this configuration within a qemu instance. We can see something like this historically done in certain SRU tests https://github.com/cloud-init/ubuntu-sru/blob/main/bugs/1682871/btest-launch.sh

Reply:
Yes, good idea

Reply:
Added in most recent push